File: impl/ocaml/src/gitignore.ml
(* A small .gitignore matcher for source auto-detection (SPEC §13.2).

   Supports blank lines, `#` comments, `!` negation, trailing `/` for
   directories, anchoring with `/`, and the `*`, `**`, and `?` wildcards. *)

open Utils

type rule = {
  pattern : Str.regexp;
  negated : bool;
  directory_only : bool;
  (* Anchored patterns match the full path relative to the ignore file. *)
  anchored : bool;
}

type t = { rules : rule list }

let glob_to_regexp glob =
  let b = Buffer.create 64 in
  Buffer.add_string b "^";
  let n = String.length glob in
  let i = ref 0 in
  while !i < n do
    (match glob.[!i] with
    | '*' ->
        if !i + 1 < n && glob.[!i + 1] = '*' then begin
          incr i;
          if !i + 1 < n && glob.[!i + 1] = '/' then (incr i; Buffer.add_string b "\\(.*/\\)?") else Buffer.add_string b ".*"
        end
        else Buffer.add_string b "[^/]*"
    | '?' -> Buffer.add_string b "[^/]"
    | '\\' when !i + 1 < n ->
        incr i;
        Buffer.add_string b (Str.quote (String.make 1 glob.[!i]))
    | c -> Buffer.add_string b (Str.quote (String.make 1 c)));
    incr i
  done;
  Buffer.add_string b "$";
  Str.regexp (Buffer.contents b)

let trim_unescaped_trailing_space line =
  let n = String.length line in
  let e = ref n in
  while !e > 0 && (line.[!e - 1] = ' ' || line.[!e - 1] = '\t') do decr e done;
  if !e > 0 && !e < n && line.[!e - 1] = '\\' then incr e;
  String.sub line 0 !e

(* Parses the content of a .gitignore file. *)
let parse content =
  let rules =
    List.filter_map
      (fun line ->
        let line = trim_suffix line "\r" in
        if String.trim line = "" || has_prefix line "#" then None
        else begin
          let line = trim_unescaped_trailing_space line in
          let negated, line =
            if has_prefix line "!" then (true, after line 1)
            else if has_prefix line "\\!" || has_prefix line "\\#" then (false, after line 1)
            else (false, line)
          in
          let directory_only, line = if has_suffix line "/" then (true, trim_suffix line "/") else (false, line) in
          let anchored, line =
            if has_prefix line "/" then (true, after line 1) else (String.contains line '/', line)
          in
          if line = "" then None else Some { pattern = glob_to_regexp line; negated; directory_only; anchored }
        end)
      (String.split_on_char '\n' content)
  in
  { rules }

let size g = List.length g.rules

(* Whether [relative_path] (relative to the directory of the ignore file,
   using `/` separators) is ignored. *)
let ignores g relative_path is_directory =
  let basename = match String.rindex_opt relative_path '/' with Some i -> after relative_path (i + 1) | None -> relative_path in
  List.fold_left
    (fun ignored rule ->
      if rule.directory_only && not is_directory then ignored
      else
        let matched = Str.string_match rule.pattern (if rule.anchored then relative_path else basename) 0 in
        if matched then not rule.negated else ignored)
    false g.rules

File: impl/go/gitignore.go
package twill

import (
	"regexp"
	"strings"
)

// A small .gitignore matcher for source auto-detection (SPEC §13.2).
//
// Supports blank lines, `#` comments, `!` negation, trailing `/` for
// directories, anchoring with `/`, and the `*`, `**`, and `?` wildcards.

type gitignoreRule struct {
	pattern       *regexp.Regexp
	negated       bool
	directoryOnly bool
	// Anchored patterns match the full path relative to the ignore file.
	anchored bool
}

// Gitignore is a parsed .gitignore file.
type Gitignore struct {
	rules []gitignoreRule
}

func gitignoreGlobToRegexp(glob string) *regexp.Regexp {
	var b strings.Builder
	b.WriteString("^")
	for i := 0; i < len(glob); i++ {
		c := glob[i]
		switch {
		case c == '*':
			if i+1 < len(glob) && glob[i+1] == '*' {
				i++
				if i+1 < len(glob) && glob[i+1] == '/' {
					i++
					b.WriteString("(?:.*/)?")
				} else {
					b.WriteString(".*")
				}
			} else {
				b.WriteString("[^/]*")
			}
		case c == '?':
			b.WriteString("[^/]")
		case c == '\\' && i+1 < len(glob):
			i++
			b.WriteString(regexp.QuoteMeta(string(glob[i])))
		default:
			b.WriteString(regexp.QuoteMeta(string(c)))
		}
	}
	b.WriteString("$")
	return regexp.MustCompile(b.String())
}

// ParseGitignore parses the content of a .gitignore file.
func ParseGitignore(content string) *Gitignore {
	g := &Gitignore{}
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSuffix(line, "\r")
		if strings.TrimSpace(line) == "" || strings.HasPrefix(line, "#") {
			continue
		}
		line = trimUnescapedTrailingSpace(line)
		negated := false
		if strings.HasPrefix(line, "!") {
			negated = true
			line = line[1:]
		} else if strings.HasPrefix(line, `\!`) || strings.HasPrefix(line, `\#`) {
			line = line[1:]
		}
		directoryOnly := false
		if strings.HasSuffix(line, "/") {
			directoryOnly = true
			line = line[:len(line)-1]
		}
		anchored := false
		if strings.HasPrefix(line, "/") {
			anchored = true
			line = line[1:]
		} else if strings.Contains(line, "/") {
			anchored = true
		}
		if line == "" {
			continue
		}
		g.rules = append(g.rules, gitignoreRule{
			pattern:       gitignoreGlobToRegexp(line),
			negated:       negated,
			directoryOnly: directoryOnly,
			anchored:      anchored,
		})
	}
	return g
}

func trimUnescapedTrailingSpace(line string) string {
	end := len(line)
	for end > 0 && (line[end-1] == ' ' || line[end-1] == '\t') {
		end--
	}
	if end > 0 && end < len(line) && line[end-1] == '\\' {
		end++
	}
	return line[:end]
}

// Size is the number of rules.
func (g *Gitignore) Size() int { return len(g.rules) }

// Ignores reports whether relativePath (relative to the directory of the
// ignore file, using `/` separators) is ignored.
func (g *Gitignore) Ignores(relativePath string, isDirectory bool) bool {
	ignored := false
	basename := relativePath[strings.LastIndex(relativePath, "/")+1:]
	for _, rule := range g.rules {
		if rule.directoryOnly && !isDirectory {
			continue
		}
		var matched bool
		if rule.anchored {
			matched = rule.pattern.MatchString(relativePath)
		} else {
			matched = rule.pattern.MatchString(basename)
		}
		if matched {
			ignored = !rule.negated
		}
	}
	return ignored
}
